frame/vamm/src/helpers/math.rs

```diff
@@ -0,0 +1,122 @@
+use crate::{Config, Error, Pallet};
+use frame_support::pallet_prelude::*;
+use sp_core::U256;
+use sp_runtime::{traits::Zero, ArithmeticError, FixedPointNumber};
+
+impl<T: Config> Pallet<T> {
+	/// Returns the vamm invariant (aka. `K`), given `base` and `quote` asset
+	/// amounts, where `K` can be derived as `base * quote`.
+	///
+	/// # Assumptions or Requirements
+	/// In order to compute the invariant, both `base` and `quote` amounts must
+	/// be greater than zero, as it's strictly forbidden for a vamm to have a
+	/// invariant equal zero.
+	///
+	/// # Errors
+	///
+	/// * [`Error::<T>::BaseAssetReserveIsZero`]
+	/// * [`Error::<T>::QuoteAssetReserveIsZero`]
+	/// * [`Error::<T>::FailedToDeriveInvariantFromBaseAndQuoteAsset`]
+	/// * [`Error::<T>::InvariantIsZero`]
+	/// * [`ArithmeticError`](sp_runtime::ArithmeticError)
+	pub fn compute_invariant(base: T::Balance, quote: T::Balance) -> Result<U256, DispatchError> {
+		// Neither base nor quote asset are allowed to be zero since it
+		// would mean the invariant would also be zero.
+		ensure!(!base.is_zero(), Error::<T>::BaseAssetReserveIsZero);
+		ensure!(!quote.is_zero(), Error::<T>::QuoteAssetReserveIsZero);
+
+		let base_u256 = Self::balance_to_u256(base)?;
+		let quote_u256 = Self::balance_to_u256(quote)?;
+		let invariant = base_u256
+			.checked_mul(quote_u256)
+			.ok_or(Error::<T>::FailedToDeriveInvariantFromBaseAndQuoteAsset)?;
+
+		ensure!(!invariant.is_zero(), Error::<T>::InvariantIsZero);
+
+		Ok(invariant)
+	}
+
+	/// Performs an exponential moving average (EMA) calculation following the formula:
```

Comment:
NIT: I think "Calculates the exponential moving average (EMA) following the formula" sounds more natural